orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/artifacts/FindArtifactsFromResourceTask.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2018 Joel Wilsson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.netflix.spinnaker.orca.clouddriver.tasks.artifacts;
+
+import com.netflix.spinnaker.kork.artifacts.model.Artifact;
+import com.netflix.spinnaker.kork.artifacts.model.ExpectedArtifact;
+import com.netflix.spinnaker.orca.ExecutionStatus;
+import com.netflix.spinnaker.orca.Task;
+import com.netflix.spinnaker.orca.TaskResult;
+import com.netflix.spinnaker.orca.clouddriver.OortService;
+import com.netflix.spinnaker.orca.clouddriver.model.Manifest;
+import com.netflix.spinnaker.orca.clouddriver.tasks.AbstractCloudProviderAwareTask;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import com.netflix.spinnaker.orca.pipeline.util.ArtifactResolver;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.Nonnull;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Slf4j
+@Component
+public class FindArtifactsFromResourceTask extends AbstractCloudProviderAwareTask implements Task {
+  public static final String TASK_NAME = "findArtifactsFromResource";
+
+  @Autowired
+  ArtifactResolver artifactResolver;
+
+  @Autowired
+  OortService oortService;
+
+  @Nonnull
+  @Override
+  public TaskResult execute(@Nonnull Stage stage) {
+    StageData stageData = stage.mapTo(StageData.class);
+    String account = getCredentials(stage);
+    Set<String> expectedArtifactIds = new HashSet<>(stageData.expectedArtifactIds);
+    List<ExpectedArtifact> expectedArtifacts = artifactResolver.getExpectedArtifacts(stage);
+    Map<String, Object> outputs = new HashMap<>();
+
+    List<ExpectedArtifact> unresolvedExpectedArtifacts = expectedArtifacts.stream()
+      .filter(a -> expectedArtifactIds.contains(a.getId()))
+      .filter(a -> a.getBoundArtifact() == null
```

Comment:
I don't think this check is necessary - the "boundArtifact" should never be null after a stage has executed.

This means that any expected artifacts you found in upstream stages should already be bound, and the ones declared in this stage need to be bound.

Reply:
This is likely to often be the first stage after Configuration. Isn't it possible for "boundArtifact" to be null, then?

For example, a pipeline triggered by a GitHub webhook after a manifest has been updated, but where the Docker image being an expected artifact defined in Configuration has no default (and doesn't use the prior execution either)?

That's the use-case I have in mind.

Reply:
As discussed on Slack, this (updating the image from a Docker trigger) should just be a separate pipeline.